app/controllers/api/base_controller/renderer.rb

```diff
@@ -514,7 +514,10 @@ def api_user_role_allows?(action_identifier)
 
         return custom_api_user_role_allows?(action_identifier) if custom_api_user_role_allows_method?(action_identifier)
 
-        Array(action_identifier).any? { |identifier| User.current_user.role_allows?(:identifier => identifier) }
+        Array(action_identifier).any? do |identifier|
+          @role_allows_cache ||= {}
+          @role_allows_cache[identifier] ||= User.current_user.role_allows?(:identifier => identifier)
```

Comment:
Won’t it re-evaluate all the ones that resulted in `false`?  The any? loop will stop on the first one returning `true`, I believe.  So it may be better to just remember all the identifiers, because if it seen again, it must mean that they evaluated to `false` before.

Reply:
Thinking further, is the following any better?
`Array(action_identifier).uniq.any? { |identifier| User.current_user.role_allows?(:identifier => identifier) }`

Reply:
> Won’t it re-evaluate all the ones that resulted in `false`?

Good catch, I can refactor to not use `||=` and instead use something like:

```ruby
unless @role_allows_cache.key?(identifier)
            @role_allows_cache[identifier] ||= User.current_user.role_allows?(:identifier => identifier)
end
@role_allows_cache[identifier]
```

I should also add some specs for this as well.

> Thinking further, is the following any better?

The problem is that this is done inside an N+1, or for each record, so the `uniq` serves no purpose here since they are all `unique` to begin with.

Effectively, we are doing this:

```ruby
# heavily simplifying to showcase the problem
@vms        = Vms.all
@reponse = []
@vms.each do |vm|
  @response << vm.to_json
  @response[:actions] = []
  fetch_typed_subcollection_actions.each do |action_identifier|
    if Array(action_identifier).any? { |identifier| User.current_user.role_allows?(:identifier => identifier) }
      @response[:actions] << action_identifier
    end
  end
end
```

So the performance problem lies that we have to go through that `.any?` a bunch of times and calculate the same work each time.  `api_user_role_allows?` just happens to accept an array, but for this case, it is most likely coming from here:

https://github.com/ManageIQ/manageiq-api/blob/2506100b/app/controllers/api/base_controller/renderer.rb#L451-L469

Where it actually passes a single value and does the looping itself (like I showed above).  What this solves is trying to check the same set of role permissions multiple times per request for each object brought back, when the result from that check is always the same.

Reply:
@NickLaMuro I think the refactoring you are suggesting would be a better approach.

Reply:
Woops, copy pasta mistake on my end.  I agree, that was my intent to do that, but thanks for catching.

Reply:
This is fixed now.

> I should also add some specs for this as well.

I am actually going to pass on this, since I did some `grep`ing for "role" and it seems like this is pretty well exercised indirectly already, and I would have to add a entirely new spec folder to "unit" test the caching of this method this since our specs are all request spec based.

Unless that is desired, I will probably mark this as resolved.